pkgs/build-support/rust/default.nix

```diff
@@ -66,14 +72,126 @@ let
     else ''
       cargoDepsCopy="$sourceRoot/${cargoVendorDir}"
     '';
+  
+  targetIsJSON = stdenv.lib.hasSuffix ".json" target;
+
+  rustSrc = stdenv.mkDerivation {
+    name = "rust-src";
+    src = rustPlatform.rust.rustc.src;
+    preferLocalBuild = true;
+    phases = [ "unpackPhase" "installPhase" ];
+    installPhase = "cp -r src $out";
+  };
+
+  toTOML = value: builtins.readFile (runCommandNoCC "tmp.toml" {
+    nativeBuildInputs = [ remarshal ];
+    value = builtins.toJSON value;
+    passAsFile = [ "value" ];
+  } ''
+    json2toml "$valuePath" "$out"
+  '');
+
+  # sysroot logic based on https://github.com/DianaNites/cargo-sysroot/
+
+  sysrootSrc = let
+    parentToml = builtins.readFile (src + /Cargo.toml);
+    parentTomlParsed = builtins.fromTOML parentToml;
+    parentProfile = if builtins.hasAttr "profile" parentTomlParsed
+                    then parentTomlParsed.profile else {};
+    sysrootProfileToml = toTOML parentProfile;
+  in stdenv.mkDerivation {
+    name = "sysroot-src";
+    phases = [ "installPhase" ];
+    installPhase = ''
+      mkdir -p $out
+      cat << EOF > $out/Cargo.toml
+      [package]
+      name = "alloc"
+      version = "0.0.0"
+      authors = ["The Rust Project Developers"]
+      edition = "2018"
+      [dependencies.compiler_builtins]
+      version = "0.1.0"
+      features = ["rustc-dep-of-std", "mem"]
+
+      [dependencies.core]
+      path = "${rustSrc}/libcore"
+
+      [lib]
+      name = "alloc"
+      path = "${rustSrc}/liballoc/lib.rs"
+      [patch.crates-io.rustc-std-workspace-core]
+      path = "${rustSrc}/tools/rustc-std-workspace-core"
+
+      ${sysrootProfileToml}
```

Comment:
My recommendation here is to use some scripting language i.e. python with a toml library to build this to concatenate this.